kvserver/rangefeed: introduce RangefeedUseBufferedSender

wenyihu6 opened the pull request:
**kvserver/rangefeed: change p.Register to use registration interface**

Previously, we introduced a registration interface to abstract the
implementation details of buffered and unbuffered registration. However, we
forgot to update the p.Register function to use this interface as well. This
patch makes the changes accordingly.

Part of: https://github.com/cockroachdb/cockroach/issues/126560
Release note: none

---

**kvserver/rangefeed: introduce RangefeedUseBufferedSender**

This patch introduces a new cluster setting RangefeedUseBufferedSender. It is
used to control whether rangefeed processor should use a buffered sender or not.
When enabled, the buffered sender is used instead of buffering events separately
per client per range. It is disabled by default. Note that this is currently
unimplemented and enabling this will cause panics. We will leave this cluster
setting undocumented until this feature is complete.

Part of: https://github.com/cockroachdb/cockroach/issues/126560
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add new cluster setting RangefeedUseBufferedSender

This patch introduces a new cluster setting RangefeedUseBufferedSender. It is
used to control whether rangefeed processor should use a buffered sender or not.
When enabled, the buffered sender is used instead of buffering events separately
per client per range. It is disabled by default. Note that this is currently
unimplemented and enabling this will cause panics. We will leave this cluster
setting undocumented until this feature is complete.

Part of: cockroachdb#126560
Release note: none

wenyihu6 committed:
kvserver/rangefeed: change p.Register to use registration interface

Previously, we introduced a registration interface to abstract the
implementation details of buffered and unbuffered registration. However, we
forgot to update the p.Register function to use this interface as well. This
patch makes the changes accordingly.

Part of: cockroachdb#126560
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127878)
<!-- Reviewable:end -->

wenyihu6 committed:
register clean up

wenyihu6 committed:
no buffered

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
split

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
plumbed wrong to be squashed

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
use more interface - see later if there is a better way not a fan of overusing interface

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
start queue

wenyihu6 committed:
util/queue: add more tests

This patch cleans the test code by modularizing the code, moving the test code
to helpers_test, and abstracting runQueueTest to use a testQueueInterface.
Currently, only Queue[T] implements it. But in a future patch, a new data
structure QueueWithFixedChunkSize[T] will also implement it.

Part of: #126560
Release note: none

wenyihu6 committed:
util/queue: add QueueWithFixedChunkSize

This patch adds a new data structure QueueWithFixedChunkSize which is like Queue
but uses a fixed size for the chunked linked list. Each chunk has a fixed size
of 4096 elements. This implementation uses sync.Pool to reduce the number of
allocations. It should be used when the queue is used to store a large number of
elements.

Enqueue, Dequeue, Len, Empty run in constant time. RemoveAll runs in linear time
with respect to the number of elements in the queue. This structure is not safe
for concurrent use.

```
goarch: arm64
BenchmarkQueueWithFixedChunkSize
BenchmarkQueueWithFixedChunkSize           26445            118984 ns/op              11 B/op          0 allocs/op
BenchmarkQueue
BenchmarkQueue                             28495            124318 ns/op          239060 B/op        446 allocs/op
PASS
```

Part of: https://github.com/cockroachdb/cockroach/issues/126560
Release note: none

# Conflicts:
#	pkg/util/queue/BUILD.bazel
#	pkg/util/queue/helpers_test.go

wenyihu6 committed:
util/queue: rename purge to removeAll

This patch renames purge to removeAll since it is more descriptive of what the
function does.

Epic: none
Release note: none

wenyihu6 committed:
a

wenyihu6 committed:
more 09-10
